core/arch/riscv/kernel/thread_rv.S

```diff
@@ -0,0 +1,376 @@
+/* SPDX-License-Identifier: BSD-2-Clause */
+/*
+ * Copyright 2022-2023 NXP
+ */
+
+#include <asm.S>
+#include <generated/asm-defines.h>
+#include <keep.h>
+#include <kernel/thread.h>
+#include <kernel/thread_private.h>
+#include <mm/core_mmu.h>
+#include <riscv.h>
+#include <riscv_macros.S>
+
+.macro get_thread_ctx res, tmp0
+	lw	\tmp0, THREAD_CORE_LOCAL_CURR_THREAD(tp)
+	la	\res, threads
+1:
+	beqz	\tmp0, 2f
+	addi	\res, \res, THREAD_CTX_SIZE
+	addi	\tmp0, \tmp0, -1
+	bnez	\tmp0, 1b
+2:
+.endm
+
+.macro save_regs, mode
+	addi	sp, sp, -THREAD_TRAP_REGS_SIZE
+.if \mode == TRAP_MODE_USER
+
+	/* Save user thread pointer and load kernel thread pointer */
+	store_xregs sp, THREAD_TRAP_REG_TP, REG_TP
+	load_xregs sp, (THREAD_TRAP_REGS_SIZE - RISCV_XLEN_BYTES), REG_TP
+
+	store_xregs sp, THREAD_TRAP_REG_GP, REG_GP
+
+	/*
+	 * Set the scratch register to 0 such in case of a recursive
+	 * exception thread_trap_vect() knows that it is emitted from kernel.
+	 */
+	csrrw	gp, CSR_XSCRATCH, zero
+	store_xregs sp, THREAD_TRAP_REG_SP, REG_GP
```

Comment:
THREAD_TRAP_REG_SP should be THREAD_TRAP_REG_GP？

Reply:
@lubing521 No.